### PR TITLE
Improve implementation of analyzer loader

### PR DESCRIPTION
**Pull request by tmat**
- Avoid holding a lock while calling the CLR loader
- unify Desktop and Core CLR implementations
- avoid dependency on the order of analyzers on the command line when compiling via csc/vbc

Fixes deadlock when bootstraping Roslyn on Linux with CoreFX rc4-24211-01

Note: VSI tests are expected to fail atm due to some cross-repo dependencies. 
Need to merge https://github.com/dotnet/roslyn-internal/pull/1062 first.

I have filed issue https://github.com/dotnet/roslyn/issues/11977 to add more test coverage and https://github.com/dotnet/roslyn/issues/11978 to suppress warnings on Linux reported by analyzers once this is change is merged.

### CODE REVIEW

**Comment by tmat** [comment]
@mavasani Please review
@srivatsn @ManishJayaswal @tmeschter FYI

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    internal abstract class AnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
+    {
+        private readonly object _guard = new object();
+
+        // lock _guard to read/write
+        private readonly Dictionary<string, Assembly> _loadedAssembliesByPath = new Dictionary<string, Assembly>();
+        private readonly Dictionary<AssemblyIdentity, Assembly> _loadedAssembliesByIdentity = new Dictionary<AssemblyIdentity, Assembly>();
+
+        // maps file name to a full path (lock _guard to read/write):
+        private readonly Dictionary<string, List<string>> _knownAssemblyPathsBySimpleName = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: `Dictionary<string, HashSet<string>>`?

Comment (tmat) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: That would be unnecessary imo. Usually there won't be many analyzers with the same simple assembly name. Rahter there will usually be one.

File: src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs [hunk 2]
```diff
@@ -75,13 +75,7 @@ private void DeleteLeftoverDirectories()
             }
         }
 
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2001:AvoidCallingProblematicMethods",
```
Comment (mavasani) [inline src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs]: We should move this suppression into the DesktopAnalyzerAssemblyLoader

Comment (mavasani) [inline src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs]: We run FXCop as part of division exit criteria near each release. Tagging @shyamnamboodiripad who used to do it earlier, I am not sure who does this now.

File: src/Compilers/Shared/DesktopAnalyzerAssemblyLoader.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    /// Assemblies will only be loaded from the locations specified when the loader
+    /// is instantiated.
+    /// </summary>
+    /// <remarks>
+    /// This type is meant to be used in scenarios where it is OK for the analyzer
+    /// assemblies to be locked on disk for the lifetime of the host; for example,
+    /// csc.exe and vbc.exe. In scenarios where support for updating or deleting
+    /// the analyzer on disk is required a different loader should be used.
+    /// </remarks>
+    internal class DesktopAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
```
Comment (mavasani) [inline src/Compilers/Shared/DesktopAnalyzerAssemblyLoader.cs]: Did we delete the file SimpleAnalyzerAssemblyLoader.cs? I see that DesktopAnalyzerAssemblyLoader.cs is being added, but the old one not deleted from disk. Not sure if its just GitHub now showing it..

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                if (!_knownAssemblyPathsBySimpleName.TryGetValue(requestedIdentity.Name, out pathList))
+                {
+                    return null;
+                }
+
+                Debug.Assert(pathList.Count > 0);
+                candidatePaths = pathList.ToImmutableArray();
+            }
+
+            // Multiple assemblies of the same simple name but different identiries might have been registered.
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Typo "identiries"

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            // Load the one that matches the requested identity (if any).
+            foreach (var candidatePath in candidatePaths)
+            {
+                var candidateIdentity = AssemblyIdentityUtils.TryGetAssemblyIdentity(candidatePath);
+
+                if (requestedIdentity.Equals(candidateIdentity))
+                {
+                    var assembly = LoadFromPathUnchecked(candidatePath);
+                    return AddToCache(assembly, requestedIdentity, candidatePath);
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: AddToCache seems unnecessary here - LoadFromPathUnchecked already does the caching.

**Comment by mavasani** [comment]
👍 Thanks!

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs (hunk 5 above)
Comment (tmat) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Good catch. Will remove

File: src/Compilers/Shared/DesktopAnalyzerAssemblyLoader.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/Shared/DesktopAnalyzerAssemblyLoader.cs]: Yes, it should be deleted. I'll double check.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Fixed.

**Comment by tmat** [comment]
@dotnet-bot retest windows_eta_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_eta_open_prtest/511/
// Retest reason:

**Comment by tmat** [comment]
test vsi please

**Comment by tmat** [comment]
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p2_open_prtest/140/
// Retest reason: ibcmerge failure

**Comment by tmat** [comment]
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p2_open_prtest/141/
// Retest reason:

**Comment by tmat** [comment]
test vsi please

**Comment by tmat** [comment]
@mavasani  The failing integration test passes on my machine.

**Comment by tmat** [comment]
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p2_open_prtest/146/
// Retest reason:

**Pull request merged**